wrappers/unity/Assets/RealSenseSDK2.0/Scripts/PointCloudGenerator.cs

```diff
@@ -10,134 +10,133 @@
+                    colorFrameHeight = frames.ColorFrame.Height;
+                    var newSize = frames.ColorFrame.Stride * colorFrameHeight;
+                    lock (l)
+                    {
+                        if (lastColorImage == null || lastColorImage.Length != newSize)
+                            lastColorImage = new byte[newSize];
+
+                        frames.ColorFrame.CopyTo(lastColorImage);
+                    }
                 }
             }
-        }
-        else if(frame.Profile.Stream == Intel.RealSense.Stream.Color)
-        {
-            //pc.MapTexture(frame);
+            pointsQueue.Enqueue(points);
         }
     }
 
-    private void UpdateParticleParams(int width, int height, Format format)
+    private bool UpdateParticleParams(int width, int height)
     {
-        streamWidth = width;
-        streamHeight = height;
-
-        if (format != Format.Z16)
+        var numParticles = (width * height);
+        if (particles.Length != numParticles)
         {
-            Debug.Log("Unsupported format");
-            return;
+            particles = new ParticleSystem.Particle[numParticles];
         }
 
-        const int bpp = 2;
-
-        if (totalImageSize != streamWidth * streamHeight * bpp)
-        {
-            totalImageSize = streamWidth * streamHeight * bpp;
-            particleSize = totalImageSize / skipParticles;
-            particles = new ParticleSystem.Particle[particleSize];
-        }
-
-        if (particleSize != totalImageSize / skipParticles)
-        {
-            particleSize = totalImageSize / skipParticles;
-            particles = new ParticleSystem.Particle[particleSize];
-        }
-        particleCount = particleSize;
+        return true;
     }
 
     void Update()
     {
-        //TODO: Lock & copy particles?
-        pointCloudParticles.SetParticles(particles, particleCount);
-    }
-
-    private bool InitializeStream(PipelineProfile activeProfile)
-    {
-        var depthStream = activeProfile.Streams.FirstOrDefault(s => s.Stream == Intel.RealSense.Stream.Depth);
-        if (depthStream == null)
+        Frame frame;
+        if (pointsQueue.PollForFrame(out frame))
         {
-            Debug.Log("No Depth stream available");
-            return false;
-        }
-        var depthProfile = depthStream as VideoStreamProfile;
-        //depthIntrinsic = depthProfile.GetIntrinsics();
-        streamWidth = depthProfile.Width;
-        streamHeight = depthProfile.Height;
-        return true;
-    }
+            using (Points points = frame as Points)
+            {
+                if (points == null)
+                    throw new Exception("Frame in queue is not a points frame");
 
-    private float Remap(float value, float from1, float to1, float from2, float to2)
-    {
-        return (value - from1) / (to1 - from1) * (to2 - from2) + from2;
+                vertices = vertices ?? new Points.Vertex[points.Count];
```

Comment:
Do not use this for Unity. `Vertex `matches `Vector3`, `TextureCoordinate `matches `Vector2`.
Use a `Vector3[]` instead of `Vertex[]` and manually marshal the array.

Then [PointCloudGenerator.cs#L126](https://github.com/zivsha/librealsense/blob/f52ced9eb9cea497cb909ec659201632014cbc58/wrappers/unity/Assets/RealSenseSDK2.0/Scripts/PointCloudGenerator.cs#L126) wouldn't allocate so much.
You're also allocating here: [PointCloudGenerator.cs#L132](https://github.com/zivsha/librealsense/blob/f52ced9eb9cea497cb909ec659201632014cbc58/wrappers/unity/Assets/RealSenseSDK2.0/Scripts/PointCloudGenerator.cs#L132), that's what `Vector3.Zero` is for...

Reply:
Sounds right, how much do you think this affects performance though?

Reply:
One less copy per vertex should help.
Still, there's more to do, particles aren't really the most performant way of rendering point clouds... Meshes are better suited, since you can update all of `Mesh.vertices` at once with one write (assuming you already have the pointcloud as a `Vector3[]`)

Reply:
@ev-mp , I've removed the redundant allocation [PointCloudGenerator.cs#L132](https://github.com/zivsha/librealsense/blob/f52ced9eb9cea497cb909ec659201632014cbc58/wrappers/unity/Assets/RealSenseSDK2.0/Scripts/PointCloudGenerator.cs#L132), but the vertex copy still requires change.